src/websocket_stream.cpp

```diff
@@ -277,6 +286,35 @@ void websocket_stream::on_close(error_code)
 	COMPLETE_ASYNC("websocket_stream::on_close");
 }
 
+void websocket_stream::on_keepalive(error_code ec)
+{
+	if (ec) return;
+
+	ADD_OUTSTANDING_ASYNC("websocket_stream::on_ping");
+	std::visit([&](auto& stream)
+		{
+			stream.async_ping({}, std::bind(&websocket_stream::on_ping
+				, shared_from_this(), _1));
+		}
+		, m_stream);
+
+}
+
+void websocket_stream::on_ping(error_code ec)
+{
+	COMPLETE_ASYNC("websocket_stream::on_ping");
+
+	if (ec) return;
+
+	arm_keepalive();
```

Comment:
This looks suspicious to me. Just because the other end is sending me a ping, does that mean I don't have to send one? This seems similar to calling `arm_keepalive()` in `on_read()`

Reply:
Yes, the naming is a bit misleading, but other callbacks follow the same naming pattern.